ID-90 A basic DrsHub README (#35)
DataBiosphere/terra-drs-hub: README.md

> # Dr. Shub, MD (Also known as DrsHub)

Reviewer: I like the fun name just as much as the next person, but I do think at the top of the readme, we have to lead with the "real" name first.

```suggestion
# DrsHub (Also known as Dr. Shub, MD)
```

And while we are here, do we know what the capitalization is supposed to be?  DRSHub or DrsHub?  I have been writing DRSHub.  Again, the top of the README is going to be the canonical source for people to reference in the future, so getting it right here will matter.  

https://broadinstitute.slack.com/archives/C03KUAJQBC2/p1657719715065769

## Overview
DrsHub is a Java Spring Boot rewrite of the Cloud Function [Martha](https://github.com/broadinstitute/martha), specifically, its v3 API.

> The purpose of DrsHub is to resolve [DRS](https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/) URLs for Terra services.

Reviewer: ```suggestion
DrsHub is the [DRS](https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/) resolution service for Terra.  

DrsHub is a Java Spring Boot rewrite of the deprecated Cloud Function [Martha](https://github.com/broadinstitute/martha), specifically, its v3 API.
```

Reasons for suggested changes.  Lead with the most important information, what DrsHub is.  Make it as succinct as possible.  The fact that this is ported from Martha is useful, but in a couple of months, nobody will care about that fact anymore.

## DRS Providers
Currently, DrsHub supports the following DRS Providers:

- Terra Data Repo (TDR)
  - Authentication Types:
    - Bearer Token
- NHGRI Analysis Visualization and Informatics Lab-space (The AnVIL)
  - Authentication Types:
    - Fence Token
- BioData Catalyst (BDC)
  - Authentication Types:
    - Fence Token
- NCI Cancer Research / Proteomics Data Commons (CRDC / PDC)
  - Authentication Types:
    - Fence Token
- Gabriella Miller Kids First DRC
  - Authentication Types:
    - Fence Token
- Passport Test Provider
  - Authentication Types:
    - Passports
    - Fence Tokens

## Usage

> To resolve a DRS URL, perform an HTTP `POST` to the `/api/v4/drs/resolve`.

Reviewer: ```suggestion
To resolve a DRS URL, perform an HTTP `POST` to `/api/v4/drs/resolve`.
```

> The content-type of your request should be application/json with the content/body of your request encoded accordingly.

Reviewer: ```suggestion
The content-type of your request should be `application/json` with the content/body of your request encoded accordingly.
```

Request bodies should look like 
```json
{
  "url": "string",
  "fields": ["string"]
}
```
where `url` is the DRS URL to resolve and `fields` is any of
```text
accessUrl
bondProvider
bucket
contentType
fileName
googleServiceAccount
gsUri
hashes
localizationPath
name
size
timeCreated
timeUpdated
```

> By default, the response will include 

Reviewer: Not sure what is meant by "by default" here.  Do you mean:

```suggestion
If no `fields` are specified in the request, the response will include the following `fields` by default: 
```

```text
bucket
contentType
fileName
gsUri
hashes
localizationPath
name
size
timeCreated
timeUpdated
googleServiceAccount
```
but this can be changed by providing values in the `fields` array.

## Architecture

> DrsHub is a Java 17 Spring Boot application running in Kubernetes. As it simply resolves urls and doesn't have any state, it has no database. For developer convenience, a Swagger UI is provided.

Reviewer: Can we provide the path to the Swagger UI?

Author: I think we would make `DrsHub` a link to the Prod swagger, once it exists and is deployed.

You must use [git-secrets](https://github.com/awslabs/git-secrets). You should be doing this anyway for all of your repositories.

Before running anything, make sure to run `./render_configs <ENV>` to render secrets locally. The default `<ENV>` is `dev`, which is what you should use for local running and testing.

DrsHub uses Gradle as a build tool. Some common Gradle commands you may want to run are
```shell
./gradlew generateSwaggerCode # Generate Swagger code for models and Swagger UI
./gradlew bootRun # Run DrsHub locally (Swagger UI at localhost:8080)
./gradlew test # Run the unit tests
./gradlew jib # Build the DrsHub Docker image
```

### Run Integration Tests
DrsHub uses [TestRunner](https://github.com/DataBiosphere/terra-test-runner) to run its integration tests.
To run the integration test suite, run
```shell
./gradlew runTest --args="suites/FullIntegration.json /tmp/test-results"
```
Adding `--stacktrace` can give you more debugging information, if needed.

### Logging
By default, DrsHub will emit logs in the Stackdriver JSON format. 
To disable this behavior for local development, add `DRSHUB_LOG_APPENDER=Console-Standard` to your environment when running DrsHub.

> ## Deployment

Reviewer: Can we add instructions for how to manually perform a deployment of the latest version and/or a specific commit or branch?

Author: Unsure if that should go in the public README or the private Playbook

    - Fence Token
- Passport Test Provider
  - Authentication Types:
    - Passports
    - Fence Tokens

## Usage
To resolve a DRS URL, perform an HTTP `POST` to the `/api/v4/drs/resolve`.
The content-type of your request should be application/json with the content/body of your request encoded accordingly.

Request bodies should look like 
```json
{
  "url": "string",
  "fields": ["string"]
}
```
where `url` is the DRS URL to resolve and `fields` is any of
```text
accessUrl
bondProvider
bucket
contentType
fileName
googleServiceAccount
gsUri
hashes
localizationPath
name
size
timeCreated
timeUpdated
```

By default, the response will include 
```text
bucket
contentType
fileName
gsUri
hashes
localizationPath
name
size
timeCreated
timeUpdated
googleServiceAccount
```
but this can be changed by providing values in the `fields` array.

## Architecture
DrsHub is a Java 17 Spring Boot application running in Kubernetes. As it simply resolves urls and doesn't have any state, it has no database. For developer convenience, a Swagger UI is provided.

Some architecture diagrams can be found in [LucidChart](https://lucid.app/documents#/documents?folder_id=297026717)

## Development

### Setup

> Install Java 17 SDK from your preferred provider. A common way to install and manage different JDK versions is to use [sdkman](https://sdkman.io/). If developing in IntelliJ, you can just configure the Project SDK to use Java 17.

Reviewer 2: I don't know if this is relevant to all devs, but I also had to update the gradle version as well (which is in a different place in the settings) to get it to use Java 17 successfully, might be worth mentioning.

Author: Oh yeah, its quite annoying.

Author: Added a bit about the gradle jvm
